bootstrap/cmd/bootstrap/app/router.go

```diff
@@ -81,48 +100,223 @@ func (r *kfctlRouter) RegisterEndpoints() {
 	http.Handle("/kfctl/apps/v1alpha2/create", optionsHandler(createHandler))
 }
 
+// decodeHTTPCreateResponse is a transport/http.DecodeResponseFunc that decodes a
+// JSON-encoded CreateDeployment response from the HTTP response body. If the response has a
+// non-200 status code, we will interpret that as an error and attempt to decode
+// the specific error message from the response body. Primarily useful in a
+// client.
+func decodeHTTPCreateResponse(_ context.Context, r *http.Response) (interface{}, error) {
+	if r.StatusCode != http.StatusOK {
+		// Try to decode the error as an httpError
+		h := httpError{}
+		err := json.NewDecoder(r.Body).Decode(&h)
+		if err == nil {
+			return nil, &h
+		}
+
+		return nil, errors.New(r.Status)
+	}
+	var resp CreateResponse
+	err := json.NewDecoder(r.Body).Decode(&resp)
+	return resp, err
+}
+
+func copyURL(base *url.URL, path string) *url.URL {
+	next := *base
+	next.Path = path
+	return &next
+}
+
+// k8sName returns the unique name for the K8s resources associated with the
+// given kubeflow deployment
+func k8sName(name string, project string, zone string) (string, error) {
+	// Project, name, and zone are required because they uniquely identify the resources
+	// that will spawned to handle the request.
+	if project == "" {
+		return "", fmt.Errorf("project is required")
+	}
+
+	if name == "" {
+		return "", fmt.Errorf("name is required")
+	}
+
+	if zone == "" {
+		return "", fmt.Errorf("zone is required")
+	}
+	h := sha256.New()
+	h.Write([]byte(fmt.Sprintf("name=%s,project=%s,zone=%s", name, project, zone)))
```

Comment:
If only the version changes we'd still want the request to be processed by the same statefulset so that we can figure out what to do about the version mismatch. 

The deployments that kfctl creates (e.g. the GCP resources only depend on project, zone, name) so it won't be unique if the only thing we change is the version.

Reply:
Got it, then zone should not be part of hash input as well since name itself is unique per project.
We might want same stateful set to handle requests with different zones. So we can dedup if they have same name.

Reply:
Good point.